webhooks/virtualmachine/validation/controller_validator.go

```diff
@@ -0,0 +1,220 @@
+// © Broadcom. All Rights Reserved.
+// The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+// SPDX-License-Identifier: Apache-2.0
+
+package validation
+
+import (
+	"fmt"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha5"
+	pkgctx "github.com/vmware-tanzu/vm-operator/pkg/context"
+	pkgutil "github.com/vmware-tanzu/vm-operator/pkg/util"
+	vmopv1util "github.com/vmware-tanzu/vm-operator/pkg/util/vmopv1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+const (
+	invalidControllerBusNumberRangeFmt     = "must be between 0 and %d"
+	invalidControllerBusNumberDoesNotExist = "controller %s:%d does not exist"
+	invalidUnitNumberReserved              = "unit number %d is reserved for the %s controller itself"
+	invalidUnitNumberRangeFmt              = "unit number must be less than %d for %s controller"
+	invalidControllerCapacityFmt           = "controller %s:%d full, maxDevices: %d"
+)
+
+// validateControllerSlots validates that all volumes can be attached
+// to controllers without exceeding the maximum number of devices per
+// controller.
+// We are not trying to be extensive here. The idea is to only perform
+// basic validations to the extent we can, and fall through to the
+// controller. It is possible that the VM is changed between the
+// webhook check and actual reconciliation, so this is best effort
+// anyway. We verify:
+//   - If a controllerBusNumber is specified, it must not be invalid
+//     (beyond max allowed). and there must be an entry in
+//     spec.hardware.controllers for this controller.
+//   - verify that we don't have more devices than max allowed for
+//     each controller type.
+func (v validator) validateControllerSlots(
+	_ *pkgctx.WebhookRequestContext,
+	vm *vmopv1.VirtualMachine) field.ErrorList {
+
+	var allErrs field.ErrorList
+
+	if len(vm.Spec.Volumes) == 0 {
+		return allErrs
+	}
+
+	volumesPath := field.NewPath("spec", "volumes")
+
+	var (
+		// Build controller map from spec.hardware.controllers
```

Comment:
GoDoc comments always begin with the name of the symbol, ex. `specControllers maps a hardware controller's ID to its entry in spec.hardware.controllers`.